src/main/java/org/ibp/api/rest/labelprinting/ObservationDatasetLabelPrinting.java

```diff
@@ -0,0 +1,562 @@
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Maps;
+import org.apache.commons.lang3.StringUtils;
+import org.generationcp.commons.util.DateUtil;
+import org.generationcp.commons.util.FileUtils;
+import org.generationcp.middleware.api.inventory.study.StudyTransactionsDto;
+import org.generationcp.middleware.api.inventory.study.StudyTransactionsRequest;
+import org.generationcp.middleware.domain.dms.DatasetDTO;
+import org.generationcp.middleware.domain.etl.MeasurementVariable;
+import org.generationcp.middleware.domain.etl.StudyDetails;
+import org.generationcp.middleware.domain.inventory.manager.TransactionsSearchDto;
+import org.generationcp.middleware.domain.oms.TermId;
+import org.generationcp.middleware.domain.ontology.VariableType;
+import org.generationcp.middleware.enumeration.DatasetTypeEnum;
+import org.generationcp.middleware.manager.Season;
+import org.generationcp.middleware.manager.api.StudyDataManager;
+import org.generationcp.middleware.service.api.dataset.DatasetService;
+import org.generationcp.middleware.service.api.dataset.DatasetTypeService;
+import org.generationcp.middleware.service.api.dataset.InstanceInfomationDTO;
+import org.generationcp.middleware.service.api.dataset.ObservationUnitData;
+import org.generationcp.middleware.service.api.dataset.ObservationUnitRow;
+import org.ibp.api.domain.common.LabelPrintingStaticField;
+import org.ibp.api.java.impl.middleware.dataset.validator.DatasetValidator;
+import org.ibp.api.java.impl.middleware.inventory.study.StudyTransactionsService;
+import org.ibp.api.java.impl.middleware.study.validator.StudyValidator;
+import org.ibp.api.rest.common.FileType;
+import org.ibp.api.rest.labelprinting.domain.Field;
+import org.ibp.api.rest.labelprinting.domain.LabelType;
+import org.ibp.api.rest.labelprinting.domain.LabelsData;
+import org.ibp.api.rest.labelprinting.domain.LabelsGeneratorInput;
+import org.ibp.api.rest.labelprinting.domain.LabelsInfoInput;
+import org.ibp.api.rest.labelprinting.domain.LabelsNeededSummary;
+import org.ibp.api.rest.labelprinting.domain.LabelsNeededSummaryResponse;
+import org.ibp.api.rest.labelprinting.domain.OriginResourceMetadata;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.i18n.LocaleContextHolder;
+import org.springframework.context.support.ResourceBundleMessageSource;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.annotation.PostConstruct;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+@Component
+@Transactional
+public class ObservationDatasetLabelPrinting extends LabelPrintingStrategy {
```

Comment:
seems to me that ObservationDatasetLabelPrinting and SubObservationDatasetLabelPrinting shares a lot of code. I think it will help to have an interface with default functions to group this code and both classes could implement it. If it complicates it now, we can experiment it in a different branch

Reply:
Totally agree. I create https://ibplatform.atlassian.net/browse/IBP-3949 to continue the refactor.